File: src/robot_life/event_engine/temporal_event_layer.py
from __future__ import annotations

from dataclasses import replace
from time import monotonic

from robot_life.common.payload_contracts import DetectionPayloadAccessor
from robot_life.common.schemas import StableEvent, new_id


class TemporalEventLayer:
    """Derive short-horizon temporal interaction events from stable detector events."""

    def __init__(
        self,
        *,
        gaze_hold_ttl_s: float = 2.5,
        attention_memory_ttl_s: float = 3.0,
    ) -> None:
        self.gaze_hold_ttl_s = max(0.5, float(gaze_hold_ttl_s))
        self.attention_memory_ttl_s = max(0.5, float(attention_memory_ttl_s))
        self._active_gaze_targets: dict[str, float] = {}
        self._active_attention_targets: dict[str, float] = {}

    def process(self, stable_event: StableEvent) -> list[StableEvent]:
        now = monotonic()
        self._prune(now)

        emitted = [stable_event]
        target_id = _target_id_from_payload(stable_event.payload)

        if stable_event.event_type in {"familiar_face_detected", "stranger_face_detected"} and target_id:
            self._active_attention_targets[target_id] = now

        if stable_event.event_type == "gaze_sustained_detected" and target_id:
            derived_type = "gaze_hold_start_detected"
            if target_id in self._active_gaze_targets:
                derived_type = "gaze_hold_active_detected"
            self._active_gaze_targets[target_id] = now
            self._active_attention_targets[target_id] = now
            emitted.append(self._derive(stable_event, event_type=derived_type, now=now))

        elif stable_event.event_type == "gaze_away_detected" and target_id:
            if target_id in self._active_gaze_targets:
                self._active_gaze_targets.pop(target_id, None)
                emitted.append(self._derive(stable_event, event_type="gaze_hold_end_detected", now=now))
                emitted.append(self._derive(stable_event, event_type="attention_lost_detected", now=now))

        elif stable_event.event_type == "gesture_detected":
            accessor = DetectionPayloadAccessor(stable_event.payload if isinstance(stable_event.payload, dict) else {})
            gesture_name = (accessor.gesture_name or "").strip().lower()
            raw_event_type = (accessor.raw_event_type or "").strip().lower()
            if (
                gesture_name in {"open_palm", "waving"}
                or raw_event_type in {"gesture_open_palm", "gesture_waving"}
                or "wave" in raw_event_type
            ):
                emitted.append(self._derive(stable_event, event_type="wave_detected", now=now))

        return emitted

    def snapshot(self) -> dict[str, object]:
        now = monotonic()
        self._prune(now)
        return {
            "active_gaze_targets": sorted(self._active_gaze_targets.keys()),
            "active_attention_targets": sorted(self._active_attention_targets.keys()),
        }

    def _prune(self, now: float) -> None:
        stale_gaze = [
            target_id
            for target_id, seen_at in self._active_gaze_targets.items()
            if (now - seen_at) > self.gaze_hold_ttl_s
        ]
        for target_id in stale_gaze:
            self._active_gaze_targets.pop(target_id, None)

        stale_attention = [
            target_id
            for target_id, seen_at in self._active_attention_targets.items()
            if (now - seen_at) > self.attention_memory_ttl_s
        ]
        for target_id in stale_attention:
            self._active_attention_targets.pop(target_id, None)

    @staticmethod
    def _derive(stable_event: StableEvent, *, event_type: str, now: float) -> StableEvent:
        payload = dict(stable_event.payload)
        payload["derived_from_event_type"] = stable_event.event_type
        payload["derived_temporal_event"] = event_type
        return replace(
            stable_event,
            stable_event_id=new_id(),
            event_type=event_type,
            valid_until_monotonic=max(stable_event.valid_until_monotonic, now + 0.8),
            stabilized_by=list(stable_event.stabilized_by) + ["temporal_event_layer"],
            payload=payload,
        )


def _target_id_from_payload(payload: dict) -> str | None:
    return DetectionPayloadAccessor(payload).target_id

File: Robot_Life_CPP/src/event_engine/temporal_event_layer.cpp
#include "robot_life_cpp/event_engine/temporal_event_layer.hpp"

#include <algorithm>

namespace robot_life_cpp::event_engine {

TemporalEventLayer::TemporalEventLayer(double gaze_hold_ttl_s, double attention_memory_ttl_s)
    : gaze_hold_ttl_s_(std::max(0.5, gaze_hold_ttl_s)),
      attention_memory_ttl_s_(std::max(0.5, attention_memory_ttl_s)) {}

std::vector<common::StableEvent> TemporalEventLayer::process(
    const common::StableEvent& stable_event,
    double now_mono_s) {
  prune(now_mono_s);
  std::vector<common::StableEvent> emitted{};
  emitted.push_back(stable_event);

  const auto target_id = target_id_from_payload(stable_event.payload);
  if ((stable_event.event_type == "familiar_face_detected" ||
       stable_event.event_type == "stranger_face_detected") &&
      !target_id.empty()) {
    active_attention_targets_[target_id] = now_mono_s;
  }

  if (stable_event.event_type == "gaze_sustained_detected" && !target_id.empty()) {
    std::string derived = "gaze_hold_start_detected";
    if (active_gaze_targets_.contains(target_id)) {
      derived = "gaze_hold_active_detected";
    }
    active_gaze_targets_[target_id] = now_mono_s;
    active_attention_targets_[target_id] = now_mono_s;
    emitted.push_back(derive_event(stable_event, derived, now_mono_s));
  } else if (stable_event.event_type == "gaze_away_detected" && !target_id.empty()) {
    if (active_gaze_targets_.contains(target_id)) {
      active_gaze_targets_.erase(target_id);
      emitted.push_back(derive_event(stable_event, "gaze_hold_end_detected", now_mono_s));
      emitted.push_back(derive_event(stable_event, "attention_lost_detected", now_mono_s));
    }
  } else if (stable_event.event_type == "gesture_detected") {
    const auto gesture_it = stable_event.payload.find("gesture_name");
    const auto raw_it = stable_event.payload.find("raw_event_type");
    const auto gesture = gesture_it == stable_event.payload.end() ? "" : gesture_it->second;
    const auto raw = raw_it == stable_event.payload.end() ? "" : raw_it->second;
    if (gesture == "open_palm" || gesture == "waving" ||
        raw == "gesture_open_palm" || raw == "gesture_waving" ||
        raw.find("wave") != std::string::npos) {
      emitted.push_back(derive_event(stable_event, "wave_detected", now_mono_s));
    }
  }

  return emitted;
}

TemporalLayerSnapshot TemporalEventLayer::snapshot(double now_mono_s) {
  prune(now_mono_s);
  TemporalLayerSnapshot snap{};
  snap.active_gaze_targets.reserve(active_gaze_targets_.size());
  snap.active_attention_targets.reserve(active_attention_targets_.size());
  for (const auto& [target, _] : active_gaze_targets_) {
    snap.active_gaze_targets.push_back(target);
  }
  for (const auto& [target, _] : active_attention_targets_) {
    snap.active_attention_targets.push_back(target);
  }
  std::sort(snap.active_gaze_targets.begin(), snap.active_gaze_targets.end());
  std::sort(snap.active_attention_targets.begin(), snap.active_attention_targets.end());
  return snap;
}

std::string TemporalEventLayer::target_id_from_payload(const common::Payload& payload) {
  const auto it = payload.find("target_id");
  return it == payload.end() ? "" : it->second;
}

common::StableEvent TemporalEventLayer::derive_event(
    const common::StableEvent& base,
    const std::string& event_type,
    double now_mono_s) {
  common::StableEvent derived = base;
  derived.stable_event_id = common::new_id();
  derived.event_type = event_type;
  derived.valid_until_monotonic = std::max(base.valid_until_monotonic, now_mono_s + 0.8);
  derived.stabilized_by.push_back("temporal_event_layer");
  derived.payload["derived_from_event_type"] = base.event_type;
  derived.payload["derived_temporal_event"] = event_type;
  return derived;
}

void TemporalEventLayer::prune(double now_mono_s) {
  for (auto it = active_gaze_targets_.begin(); it != active_gaze_targets_.end();) {
    if ((now_mono_s - it->second) > gaze_hold_ttl_s_) {
      it = active_gaze_targets_.erase(it);
    } else {
      ++it;
    }
  }

  for (auto it = active_attention_targets_.begin(); it != active_attention_targets_.end();) {
    if ((now_mono_s - it->second) > attention_memory_ttl_s_) {
      it = active_attention_targets_.erase(it);
    } else {
      ++it;
    }
  }
}

}  // namespace robot_life_cpp::event_engine
